dlt/destinations/impl/sink/sink.py

```diff
@@ -0,0 +1,172 @@
+    JobClientBase,
+)
+
+from dlt.destinations.impl.sink import capabilities
+from dlt.destinations.impl.sink.configuration import SinkClientConfiguration, TSinkCallable
+
+
+class SinkLoadJob(LoadJob, ABC):
+    def __init__(
+        self,
+        table: TTableSchema,
+        file_path: str,
+        config: SinkClientConfiguration,
+        schema: Schema,
+        load_package_state: Dict[str, int],
+    ) -> None:
+        super().__init__(FileStorage.get_file_name_from_file_path(file_path))
+        self._file_path = file_path
+        self._config = config
+        self._table = table
+        self._schema = schema
+
+        self._state: TLoadJobState = "running"
+        self._storage_id = f"{self._parsed_file_name.table_name}.{self._parsed_file_name.file_id}"
+        try:
+            current_index = load_package_state.get(self._storage_id, 0)
+            for batch in self.run(current_index):
+                self.call_callable_with_items(batch)
+                current_index += len(batch)
+                load_package_state[self._storage_id] = current_index
+
+            self._state = "completed"
+        except Exception as e:
+            self._state = "retry"
+            raise e
+        finally:
+            # save progress
+            commit_load_package_state()
+
+    @abstractmethod
+    def run(self, start_index: int) -> Iterable[TDataItems]:
+        pass
+
+    def call_callable_with_items(self, items: TDataItems) -> None:
+        if not items:
+            return
+
+        # coerce items into correct format specified by schema
+        coerced_items: TDataItems = []
+        for item in items:
+            coerced_item, table_update = self._schema.coerce_row(self._table["name"], None, item)
+            assert not table_update
+            coerced_items.append(coerced_item)
+
+        # send single item on batch size 1
+        if self._config.batch_size == 1:
+            coerced_items = coerced_items[0]
+
+        # call callable
+        self._config.credentials.resolved_callable(coerced_items, self._table)
+
+    def state(self) -> TLoadJobState:
+        return self._state
+
+    def exception(self) -> str:
+        raise NotImplementedError()
+
+
+class SinkParquetLoadJob(SinkLoadJob):
+    def run(self, start_index: int) -> Iterable[TDataItems]:
+        # stream items
+        from dlt.common.libs.pyarrow import pyarrow
+
+        # guard against changed batch size after restart of loadjob
+        assert (
+            start_index % self._config.batch_size
+        ) == 0, "Batch size was changed during processing of one load package"
+
+        start_batch = start_index / self._config.batch_size
+        with pyarrow.parquet.ParquetFile(self._file_path) as reader:
+            for record_batch in reader.iter_batches(batch_size=self._config.batch_size):
+                if start_batch > 0:
+                    start_batch -= 1
+                    continue
+                batch = record_batch.to_pylist()
```

Comment:
hmmmmm. IMO just yield arrow here. Alex for example is pushing panda frames to salesforce

Reply:
done